### PR TITLE
#77 Implement postconditions to replay a changeset until it's done

### PR DESCRIPTION
**Pull request by fpavageau**
I'm also updating the documentation, I'll create another PR for that shortly.

### CODE REVIEW

**Comment by fbiville** [comment]
Awesome! I cannot wait to review this tonight!!
By the way, would you mind opening a similar PR against `master_3.x` (you might face minor conflicts). This branch works with the new JDBC driver.

File: liquigraph-core/src/main/java/org/liquigraph/core/io/ChangelogGraphWriter.java [hunk 1]
```diff
@@ -80,36 +80,22 @@ public void write(Collection<Changeset> changelog) {
     }
 
     private StatementExecution executeStatement(Changeset changeset) {
-        try (Statement statement = connection.createStatement()) {
+        try {
             Precondition precondition = changeset.getPrecondition();
             boolean preconditionResult = executePrecondition(precondition);
+            // Make sure the precondition does not actually modify the data
+            connection.rollback();
```
Comment (fbiville) [inline liquigraph-core/src/main/java/org/liquigraph/core/io/ChangelogGraphWriter.java]: :+1:

File: liquigraph-core/src/main/java/org/liquigraph/core/model/Precondition.java [hunk 2]
```diff
@@ -40,6 +40,7 @@ public void setPolicy(PreconditionErrorPolicy policy) {
         @XmlElementRef(name = "or", type = OrQuery.class),
         @XmlElementRef(name = "query", type = SimpleQuery.class)
     })
+    @Override
```
Comment (fbiville) [inline liquigraph-core/src/main/java/org/liquigraph/core/model/Precondition.java]: nitpick: empty line before `@Override`? :)

**Comment by fpavageau** [comment]
Sure, I'm duplicating the branch for rebasing onto master_3.x right now, the other PR should be there shortly.

File: liquigraph-core/src/test/java/org/liquigraph/core/io/ChangelogGraphWriterTestSuite.java [hunk 3]
```diff
@@ -297,4 +339,9 @@ private static Object property(Object changeset, String name) {
         assertThat(changeset).isInstanceOf(Node.class);
         return ((Node)changeset).getProperty(name);
     }
+
+    private void given_inserted_data(String query, Connection connection) throws SQLException {
+        connection.createStatement().executeQuery(query);
+        connection.commit();
```
Comment (fpavageau) [inline liquigraph-core/src/test/java/org/liquigraph/core/io/ChangelogGraphWriterTestSuite.java]: By the way, that simple `commit()` is the fix for the failure of the unit test I had during the Hackergarten. I could have sworn I'd tested it at the time, though.

Comment (fbiville) [inline liquigraph-core/src/test/java/org/liquigraph/core/io/ChangelogGraphWriterTestSuite.java]: Wow, OK...

File: liquigraph-core/src/main/java/org/liquigraph/core/model/Precondition.java (hunk 2 above)
Comment (fpavageau) [inline liquigraph-core/src/main/java/org/liquigraph/core/model/Precondition.java]: Why? It's preceded by another annotation, not a method.

**Comment by fpavageau** [comment]
OK, so porting to `master_3.x` is going to take a bit longer, as Bolt works differently: you can't "commit, rollback, commit" on a single connection, once it has been rollbacked it cannot be used anymore and needs to be replaced.

**Comment by fbiville** [comment]
@fpavageau you're specificically talking about the rollback done at the condition execution?

**Comment by fpavageau** [comment]
Yes, that rollback messes up the process when running on Bolt. I wonder if it fails in other conditions as well, actually. I seem to remember you disabled the remote tests on Travis a while back, is it still the case? I.e. was my modification tested in remote mode?

I think I'll rework it on this branch, instead of the one ported to `master_3.x`. I realized I forgot to rollback the postcondition query the same way as the precondition.

It seems my only option is to pass the `LiquigraphJdbcConnector` to the `ChangelogWriter`, so that it can create its own connections for the pre- and postconditions and rollback them, while using the main connection to run the migrations and commit them.

Because you end up with several connections which need to see the same consistent state after each commit, I think it might be worth mentioning in the documentation that Liquigraph should be run against the master on a HA cluster with Enterprise Edition. It might have already been the case when used on a remote instance, anyway, if the JDBC driver opens multiple HTTP connections in sequence.

**Comment by fbiville** [comment]
@fpavageau you are right about hitting the same member of the cluster for the repeated execution (while it should, it doesn't have to be the main node, or?)

I need to re-enable Docker tests with Neo4j 2.3.4. I'll push a PR tonight and see how it goes.

Shall I wait for more changes before merging this PR?

**Comment by fbiville** [comment]
You can rebase on `master`, I have reinstated Docker tests and updated the latest 2.2.x and 2.3.x branches.

**Comment by fpavageau** [comment]
I've rebased on master. If the build passes, I think it can be merged. Unless you have comments, of course 😄.

**Comment by fpavageau** [comment]
OK, it works correctly now. I also fixed the tests starting too early for the dockerized Neo4j.

For me, it's ready to be merged.

**Comment by fbiville** [comment]
Great, will review and merge this today or tomorrow!

**Comment by fbiville** [comment]
Squashed in 3 commits and merged to `master`.

**Pull request closed without merging**